tests/kernel/spinlock/src/spinlock_fairness.c

```diff
@@ -0,0 +1,178 @@
+	 * at the beginning of the test.
+	 */
+	int key = arch_irq_lock();
+
+	atomic_dec(&start_sync);
+	while (atomic_get(&start_sync) != 0)
+		;
+
+	/*
+	 * Run the test: let the cores contend for the spinlock and
+	 * collect the spinlock acquisition statistics
+	 */
+	do {
+		k_spinlock_key_t spinlock_key = k_spin_lock(&lock);
+
+		if (atomic_get(&fairness_test_cycles) == 0) {
+			k_spin_unlock(&lock, spinlock_key);
+			arch_irq_unlock(key);
+			return;
+		}
+
+		spinlock_grabbed[core_id]++;
+		atomic_dec(&fairness_test_cycles);
+
+		k_spin_unlock(&lock, spinlock_key);
+	} while (atomic_get(&fairness_test_cycles) != 0);
+
+	arch_irq_unlock(key);
+}
+
+/**
+ * @brief Test initialization
+ */
+static void test_init(void)
+{
+	memset(tthread, 0x00, sizeof(tthread));
+	memset(tstack, 0x00, sizeof(tstack));
+	atomic_set(&start_sync, CORES_NUM);
+	atomic_set(&fairness_test_cycles, FAIRNESS_TEST_CYCLES_PER_CORE * CORES_NUM);
+
+	for (uintptr_t core_id = 0; core_id < CORES_NUM; core_id++) {
+		struct k_thread *thread = get_thread(core_id);
+
+		k_thread_create(thread, tstack[core_id], STACK_SIZE,
+		   (k_thread_entry_t)test_thread,
+		   (void *)core_id, NULL, NULL,
+		   K_PRIO_COOP(10), 0, K_FOREVER);
+
+		/*
+		 * Pin each thread to a particular CPU core.
+		 * The larger the core's memory access latency in comparison to the
+		 * other cores - the less chances to win a contention for the spinlock
+		 * this core will have in case the spinlock implementation doesn't
+		 * provide acquisition fairness.
+		 */
+		k_thread_cpu_pin(thread, core_id);
+	}
+}
+
+/**
+ * @brief Test spinlock acquisition fairness
+ *
+ * @details This test verifies a spinlock acquisition fairness in relation
+ *		to the cores contending for the spinlock. Memory access latency may
+ *		vary between the CPU cores, so that some CPUs reach the spinlock faster
+ *		than the others and depending on spinlock implementation may get
+ *		higher chance to win the contention for the spinlock than the other
+ *		cores, making them to starve.
+ *		This effect may become critical for some real-life platforms
+ *		(e.g. NUMA) resulting in performance loss or even a live-lock,
+ *		when a single CPU is continuously winning the contention.
+ *		This test ensures that the probability to win the contention for a
+ *		spinlock is evenly distributed between all of the contending cores.
+ *
+ * @ingroup kernel_spinlock_tests
+ *
+ * @see k_spin_lock(), k_spin_unlock()
+ */
+ZTEST(spinlock, test_spinlock_fairness)
+{
+	test_init();
+
+	/* Launching all the threads */
+	for (uintptr_t core_id = 0; core_id < CORES_NUM; core_id++) {
+		struct k_thread *thread = get_thread(core_id);
+
+		k_thread_start(thread);
+	}
+	/* Waiting for all the threads to complete */
+	for (uintptr_t core_id = 0; core_id < CORES_NUM; core_id++) {
+		struct k_thread *thread = get_thread(core_id);
+
+		k_thread_join(thread, K_FOREVER);
+	}
+
+	/* Verify spinlock acquisition fairness */
```

Comment:
It might be nice to print out the number of times  a spinlock was acquired for each core before validation